File: backend/app/db/models/market_radar.py
"""Persistent snapshots, rules, and alert events for the market radar."""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    Float,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base, TimestampMixin


class MarketRadarSnapshot(Base, TimestampMixin):
    __tablename__ = "market_radar_snapshots"
    __table_args__ = (
        CheckConstraint(
            "snapshot_type IN ('intraday', 'eod')",
            name="ck_market_radar_snapshots_snapshot_type",
        ),
        CheckConstraint(
            "status IN ('fresh', 'partial', 'stale', 'unavailable')",
            name="ck_market_radar_snapshots_status",
        ),
        CheckConstraint(
            "confidence >= 0 AND confidence <= 1",
            name="ck_market_radar_snapshots_confidence",
        ),
        Index(
            "uq_market_radar_snapshot_identity",
            "snapshot_type",
            "as_of",
            "formula_version",
            unique=True,
        ),
        Index("ix_market_radar_snapshots_type_as_of", "snapshot_type", "as_of"),
        Index("ix_market_radar_snapshots_status", "status"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    snapshot_type: Mapped[str] = mapped_column(String(16), nullable=False)
    as_of: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    computed_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    status: Mapped[str] = mapped_column(String(16), nullable=False)
    confidence: Mapped[float] = mapped_column(Float, nullable=False)
    formula_version: Mapped[str] = mapped_column(String(64), nullable=False)
    metrics_json: Mapped[str] = mapped_column(Text, nullable=False)
    source_freshness_json: Mapped[str] = mapped_column(Text, nullable=False)


class MarketAlertRule(Base, TimestampMixin):
    __tablename__ = "market_alert_rules"
    __table_args__ = (
        CheckConstraint("version >= 1", name="ck_market_alert_rules_version"),
        CheckConstraint(
            "scope IN ('market', 'sector', 'symbol', 'data')",
            name="ck_market_alert_rules_scope",
        ),
        CheckConstraint(
            "severity IN ('low', 'medium', 'high')",
            name="ck_market_alert_rules_severity",
        ),
        CheckConstraint(
            "cooldown_seconds >= 0",
            name="ck_market_alert_rules_cooldown_seconds",
        ),
        CheckConstraint(
            "source IN ('system', 'user')",
            name="ck_market_alert_rules_source",
        ),
        Index(
            "uq_market_alert_rule_key_version",
            "rule_key",
            "version",
            unique=True,
        ),
        Index(
            "ix_market_alert_rules_enabled_scope_subject",
            "enabled",
            "scope",
            "subject",
        ),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    rule_key: Mapped[str] = mapped_column(String(120), nullable=False)
    version: Mapped[int] = mapped_column(Integer, nullable=False)
    scope: Mapped[str] = mapped_column(String(16), nullable=False)
    subject: Mapped[str] = mapped_column(String(160), nullable=False)
    rule_type: Mapped[str] = mapped_column(String(80), nullable=False)
    parameters_json: Mapped[str] = mapped_column(Text, nullable=False)
    severity: Mapped[str] = mapped_column(String(16), nullable=False)
    cooldown_seconds: Mapped[int] = mapped_column(Integer, nullable=False)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)
    source: Mapped[str] = mapped_column(String(16), nullable=False)


class MarketAlertEvent(Base, TimestampMixin):
    __tablename__ = "market_alert_events"
    __table_args__ = (
        CheckConstraint(
            "scope IN ('market', 'sector', 'symbol', 'data')",
            name="ck_market_alert_events_scope",
        ),
        CheckConstraint(
            "severity IN ('low', 'medium', 'high')",
            name="ck_market_alert_events_severity",
        ),
        CheckConstraint(
            "status IN ('active', 'acknowledged', 'dismissed', 'resolved')",
            name="ck_market_alert_events_status",
        ),
        CheckConstraint(
            "occurrence_count >= 1",
            name="ck_market_alert_events_occurrence_count",
        ),
        CheckConstraint(
            "clear_streak >= 0",
            name="ck_market_alert_events_clear_streak",
        ),
        Index(
            "ux_market_alert_events_open_dedupe_key",
            "dedupe_key",
            unique=True,
            sqlite_where=text("status IN ('active', 'acknowledged', 'dismissed')"),
        ),
        Index(
            "ix_market_alert_events_status_triggered_at",
            "status",
            "triggered_at",
        ),
        Index(
            "ix_market_alert_events_scope_subject_status",
            "scope",
            "subject",
            "status",
        ),
        Index("ix_market_alert_events_rule_id", "rule_id"),
        Index("ix_market_alert_events_snapshot_id", "snapshot_id"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    rule_id: Mapped[int] = mapped_column(
        ForeignKey("market_alert_rules.id"),
        nullable=False,
    )
    snapshot_id: Mapped[int | None] = mapped_column(
        ForeignKey("market_radar_snapshots.id", ondelete="SET NULL")
    )
    scope: Mapped[str] = mapped_column(String(16), nullable=False)
    subject: Mapped[str] = mapped_column(String(160), nullable=False)
    direction: Mapped[str] = mapped_column(String(24), nullable=False)
    severity: Mapped[str] = mapped_column(String(16), nullable=False)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="active")
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    explanation: Mapped[str] = mapped_column(Text, nullable=False)
    dedupe_key: Mapped[str] = mapped_column(String(240), nullable=False)
    evidence_json: Mapped[str] = mapped_column(Text, nullable=False)
    triggered_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    last_seen_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    acknowledged_at: Mapped[datetime | None] = mapped_column(DateTime)
    dismissed_at: Mapped[datetime | None] = mapped_column(DateTime)
    resolved_at: Mapped[datetime | None] = mapped_column(DateTime)
    last_notified_at: Mapped[datetime | None] = mapped_column(DateTime)
    occurrence_count: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    clear_streak: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

File: frontend/src/api/marketRadar.ts
import request, { type RequestConfig } from './request'

export type RadarFreshness = 'fresh' | 'partial' | 'stale' | 'unavailable'
export type RadarRealtimeMode = 'push' | 'polling_30s' | 'offline' | 'closed'
export type MarketAlertStatus = 'active' | 'acknowledged' | 'dismissed' | 'resolved'
export type MarketAlertSeverity = 'low' | 'medium' | 'high'
export type MarketAlertScope = 'market' | 'sector' | 'symbol' | 'data'

export interface MarketRadarSource {
  name: string
  as_of: string | null
  status: string
  reason?: string | null
  [key: string]: unknown
}

export interface MarketRadarEnvelope<T> {
  as_of: string | null
  computed_at: string
  status: RadarFreshness
  confidence: number
  realtime_mode: RadarRealtimeMode
  sources: MarketRadarSource[]
  data: T
}

export interface MarketRadarOverview {
  alert_counts?: Record<string, number>
  emotion?: Record<string, unknown>
  crowding?: Record<string, unknown>
  [key: string]: unknown
}

export interface MarketRadarAlert {
  id: number
  rule_id: number
  snapshot_id: number | null
  scope: MarketAlertScope
  subject: string
  direction: string
  severity: MarketAlertSeverity
  status: MarketAlertStatus
  title: string
  explanation: string
  dedupe_key: string
  evidence: Record<string, unknown>
  triggered_at: string
  last_seen_at: string
  acknowledged_at: string | null
  dismissed_at: string | null
  resolved_at: string | null
  last_notified_at: string | null
  occurrence_count: number
  clear_streak: number
}

export interface MarketRadarAlertPage {
  items: MarketRadarAlert[]
  total: number
  page: number
  page_size: number
}

export type MarketRadarStreamEvent<T extends object> = T & {
  schema_version: 1
  event_id: string
  sequence: number
  occurred_at: string
}

export interface MarketRadarRule {
  id: number
  rule_key: string
  version: number
  scope: 'symbol'
  subject: string
  title: string
  direction: 'up' | 'down' | 'either'
  rule_type: 'metric_threshold'
  parameters: {
    metric: string
    operator: 'lte' | 'gte' | 'abs_gte'
    threshold: number
  }
  severity: MarketAlertSeverity
  cooldown_seconds: number
  enabled: boolean
  source: 'system' | 'user'
  created_at: string
  updated_at: string
}

export interface MarketRadarRulePage {
  items: MarketRadarRule[]
  total: number
}

export interface MarketRadarRefreshResult {
  task_id: string
  kind: 'market_radar_refresh'
  status: 'queued' | 'running'
  refresh_kind: 'intraday' | 'eod'
  trade_date: string | null
}

export type MarketRadarRuleCreate = Omit<
  MarketRadarRule,
  'id' | 'version' | 'source' | 'created_at' | 'updated_at'
>

export type MarketRadarRulePatch = Partial<Pick<
  MarketRadarRule,
  'subject' | 'title' | 'direction' | 'parameters' | 'severity' | 'cooldown_seconds' | 'enabled'
>>

interface AlertQuery {
  status?: MarketAlertStatus
  severity?: MarketAlertSeverity
  scope?: MarketAlertScope
  subject?: string
  start_at?: string
  end_at?: string
  page?: number
  page_size?: number
}

const withDate = (tradeDate?: string) => tradeDate ? { trade_date: tradeDate } : undefined

export const marketRadarApi = {
  overview: (config?: RequestConfig) =>
    request.get<MarketRadarEnvelope<MarketRadarOverview>>('/market-radar/overview', config),
  breadth: (params: { days?: number; mode?: 'percent' | 'count' } = {}) =>
    request.get<MarketRadarEnvelope<Record<string, unknown>>>('/market-radar/breadth', { params }),
  limitLadder: (tradeDate?: string) =>
    request.get<MarketRadarEnvelope<Record<string, unknown>>>('/market-radar/limit-ladder', {
      params: withDate(tradeDate),
    }),
  crowding: (params: {
    scope?: 'market' | 'sector' | 'symbol'
    subject?: string
  } = {}) => request.get<MarketRadarEnvelope<Record<string, unknown>>>(
    '/market-radar/crowding',
    { params },
  ),
  sectors: (tradeDate?: string) =>
    request.get<MarketRadarEnvelope<Record<string, unknown>>>('/market-radar/sectors', {
      params: withDate(tradeDate),
    }),
  alerts: (params: AlertQuery = {}, config?: RequestConfig) =>
    request.get<MarketRadarEnvelope<MarketRadarAlertPage>>('/market-radar/alerts', {
      ...config,
      params,
    }),
  activeHighAlerts: (config?: RequestConfig) =>
    request.get<MarketRadarEnvelope<MarketRadarAlertPage>>('/market-radar/alerts', {
      ...config,
      params: { status: 'active', severity: 'high', page: 1, page_size: 100 },
      notifyError: false,
    }),
  alert: (eventId: number) =>
    request.get<MarketRadarEnvelope<MarketRadarAlert>>(`/market-radar/alerts/${eventId}`),
  acknowledgeAlert: (eventId: number) =>
    request.post<MarketRadarEnvelope<MarketRadarAlert>>(
      `/market-radar/alerts/${eventId}/acknowledge`,
    ),
  dismissAlert: (eventId: number) =>
    request.post<MarketRadarEnvelope<MarketRadarAlert>>(`/market-radar/alerts/${eventId}/dismiss`),
  rules: (includeDisabled = false) =>
    request.get<MarketRadarEnvelope<MarketRadarRulePage>>('/market-radar/rules', {
      params: { include_disabled: includeDisabled },
    }),
  createRule: (payload: MarketRadarRuleCreate) =>
    request.post<MarketRadarEnvelope<MarketRadarRule>>('/market-radar/rules', payload),
  patchRule: (ruleId: number, payload: MarketRadarRulePatch) =>
    request.patch<MarketRadarEnvelope<MarketRadarRule>>(`/market-radar/rules/${ruleId}`, payload),
  deleteRule: (ruleId: number) =>
    request.delete<MarketRadarEnvelope<MarketRadarRule>>(`/market-radar/rules/${ruleId}`),
  refresh: (payload: { kind: 'intraday'; trade_date?: never } | { kind: 'eod'; trade_date: string }) =>
    request.post<MarketRadarEnvelope<MarketRadarRefreshResult>>('/market-radar/refresh', payload),
  streamUrl: () => '/api/market-radar/stream',
}
